src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
 
     stop_execution_clock();
 
-    if (success) {
+    if (txid.has_value()) {
         set_state(OperationStatus::SUCCESS);
     } else {
         set_state(OperationStatus::FAILED);
     }
 
     std::string s = strprintf("%s: z_sendmany finished (status=%s", getId(), getStateAsString());
-    if (success) {
-        s += strprintf(", txid=%s)\n", tx_.GetHash().ToString());
+    if (txid.has_value()) {
+        s += strprintf(", txid=%s)\n", txid.value().ToString());
     } else {
         s += strprintf(", error=%s)\n", getErrorMessage());
     }
     LogPrintf("%s",s);
-
-    // !!! Payment disclosure START
-    if (success && paymentDisclosureMode && paymentDisclosureData_.size()>0) {
-        uint256 txidhash = tx_.GetHash();
-        std::shared_ptr<PaymentDisclosureDB> db = PaymentDisclosureDB::sharedInstance();
-        for (PaymentDisclosureKeyInfo p : paymentDisclosureData_) {
-            p.first.hash = txidhash;
-            if (!db->Put(p.first, p.second)) {
-                LogPrint("paymentdisclosure", "%s: Payment Disclosure: Error writing entry to database for key %s\n", getId(), p.first.ToString());
-            } else {
-                LogPrint("paymentdisclosure", "%s: Payment Disclosure: Successfully added entry to database for key %s\n", getId(), p.first.ToString());
-            }
-        }
-    }
-    // !!! Payment disclosure END
 }
 
-struct TxValues {
+class TxValues {
+public:
     CAmount t_inputs_total{0};
     CAmount z_inputs_total{0};
     CAmount t_outputs_total{0};
     CAmount z_outputs_total{0};
     CAmount targetAmount{0};
-    bool selectedUTXOCoinbase{false};
+
+    CAmount GetChangeAmount() {
+        return t_inputs_total + z_inputs_total - targetAmount;
+    }
 };
 
+bool IsFromAnyTaddr(const PaymentSource& paymentSource) {
+    return std::visit(match {
+        [&](const FromAnyTaddr& fromAny) {
+            return true;
+        },
+        [&](const PaymentAddress& addr) {
+            return false;
+        }
+    }, paymentSource);
+}
+
+// Construct and send the transaction, returning the resulting txid.
+// Errors in transaction construction will throw.
```

Comment:
```suggestion
// Construct and send the transaction, returning the resulting txid if successful.
// Errors in transaction construction will either throw or return `std::nullopt`.
```
(if this is correct; I'm not sure).